### PR TITLE
Add a top-level `concretizer` config scope and `--fresh` option

### PR DESCRIPTION
**Pull request by tgamblin**
We want to make `--reuse` the default concretization strategy, but to do that we need to have some configuration so that users can revert if they need to.  Also, the concretizer is going to grow to have many more options, and we really need some structured place to specify them.

We currently have a couple places where concretizer options are specified (aside from the preferences in `packages.yaml`):
1. The `config:concretizer` option. This chooses the solver (`clingo` or `original`).
2. The `concretization` option in environments. This is not a top-level config section -- it's just for environments.

Extending either of these is awkward -- both are simple strings and we'd need to deprecate (or work around) the old config options to expand them into more extensive YAML options.  Also, we'd eventually like to have one place where concretizer configuration is specified.

To avoid overlapping with either of these and to allow the most extensibility in the future, this adds a new `concretizer` config section that can be used in and outside of environments. There are two options right now: `reuse` and `minimal`.  This can expand later.  My expectation is that we will soon deprecate `config:concretizer` and warn when the user doesn't use `clingo`, and we will eventually (sometime later) move the `together` / `separately` options from `concretization` into the top-level `concretizer` section.

`concretizer.yaml` currently looks like this:

```yaml
concretizer:
  # Whether to consider installed packages or packages from buildcaches when
  # concretizing specs. If `true`, we'll try to use as many installs/binaries
  # as possible, rather than building. If `false`, we'll always give you a fresh
  # concretization.
  reuse: false
```

Stay tuned for additional options like `--minimal`.

And there are *two* options on `spack install`, `spack spec`, etc.:

```
concretizer arguments:
  --fresh               do not reuse installed deps; build newest configuration
  --reuse               reuse installed dependencies/buildcaches when possible
```

`--fresh` is there for when we make `--reuse` the default.  I could not think of a more intuitive way to express what Spack's default install routine, but here is the rationale -- see if you like it:

1. Most package managers would call Spack's current default an "upgrade" install -- it prefers the latest versions.  However, it's more than that -- it looks at whatever preferences you've set up, and it may change variants and things based on what's in the repo.  So `--upgrade` didn't fully capture it (you might be "side-grading" or just re-concretizing).
2. The default install tends to create new builds.  I thought about `--new` for "new installation", but it might also reuse if there's a hash match, so `--new` didn't seem completely right.
3. People who do not like `--reuse`  -- e.g., people with CI use cases or nix/guix purists -- would call the default Spack install "deterministic", because it doesn't consider installation state.  It's reproducible for the same spack version and packages.  `--deterministic` and `--reproducible` seem too technical.

So I settled on `--fresh`, because I think it captures what we're doing -- making a fresh install from the latest packages.  It also gives us a term we can use pretty easily.

When we make `--reuse` the default, it'll make more sense, as Spack will tend to reuse what you already have, and you may want to start with a completely `--fresh` install once those installations become too old.

There are a few additional technical refactors in here to make this stuff easier; the details of those are in the commit messages.

- [x] refactor `spack.solver.asp.solve()` into a class
- [x] add a `concretizer` config section and default `concretizer.yaml`
- [x] fix up how we handle mock configuration in `conftest.py`
- [x] introduce a `ConfigSetAction` to automatically set config options with `argparse`
- [x] introduce `add_concretizer_args()` method to make concretizer arg handling consistent across commands
- [x] rework call chain from commands -> solver to reduce the number of arguments and rely on config
- [x] introduce `--fresh` concretizer arg
- [x] docs

### CODE REVIEW

**Comment by adamjstewart** [comment]
I actually prefer `--deterministic` or `--reproducible` over `--fresh`. Like you said, the people who will use this option are those who care about determinism/reproducibility. `--fresh` is kinda ambiguous much like `--dirty`.

Looking at this and #28470, I think it would be helpful to formulate as many possible concretization use cases as we possibly can. As of now, it's unclear to me if it makes sense to list these as booleans keys (i.e. some of these may not be compatible, or may not be binary). Here are some concretization schemes I think would be useful to have (names TBD):

### Determinism

* **reuse none**: spec does not depend on what is already installed (old behavior)
* **reuse all**: reuse already installed packages whenever possible (new behavior)
* **reuse deps-only**: reuse dependencies, but upgrade root specs (my desired behavior)
* **pure chaos**: make all decisions randomly, good test of dependency constraints 😈 

I think "reuse all" is a good default here.

### Packages

* **minimal builds all**: spec with the fewest nodes (useful for short build time)
* **minimal builds source**: spec with fewest builds from source, buildcache installs don't matter
* **maximal builds all**: spec with the most nodes (useful to stress test spack)
* **maximal builds source**: spec with most builds from source

### Versions

* **maximal versions all**: install the newest versions of all packages (old behavior)
* **maximal versions root**: install the newest versions of root specs, reuse installed (my desired behavior)
* **minimal versions all**: install the oldest versions of all packages (useful for CI)
* **minimal versions root**: install the oldest versions of root specs, newest version of deps (useful for CI)

See https://github.com/pypa/pip/issues/8085 for some use cases for and debate on the minimal versions concretization option in pip. I would actually love this to ensure that my documented minimal supported versions remain up-to-date. This would also help to test Spack packages.

### Compilers

* **single compiler**: current behavior to minimize compiler conflict
* **reuse compiler**: allow mixing and matching compilers or mixing and matching compiler flags to minimize rebuilds (dangerous)
* **system compiler**: use system compiler for build deps, default/specified compiler for everything else (separate concretization of build deps PR)

### Variants

* **minimal variants**: disable all boolean variants, none value in multi-valued variants (useful for testing that deps are correct)
* **maximal variants**: enable all boolean variants, all values in multi-valued variants (useful for populating buildcache with most powerful binaries)

For minimal variants, imagine just how many packages would break if you turned off all of the Python variants, we already hit this with packages like OpenCV/boost. For maximal variants, imagine distributing PyTorch/TF binaries built for every supported cuda_arch.

### Architecture

* **optimized**: only use binaries built for this specific microarchitecture
* **unoptimized**: allow unoptimized binaries to avoid unnecessary builds

If you can think of any other possible use cases, feel free to update this comment. 

Seeing these all laid out, it's unclear if it makes sense to have separate flags/keys for all of these or just a few flags/keys like `--concretization-strategy=reuse-root`. Of course, we don't necessarily have to support all of these, but I've already mentioned at least a few that I personally want that we don't yet have, and I'm sure others have more specific use cases.

**Comment by tgamblin** [comment]
@adamjstewart: I actually like all of these.  They're all doable.  Some of them (like compilers) require some information that we don't have yet (specifically, compatibility between compilers and compiler versions for different languages).  It might be worth converting this list to a discussion someplace so we can keep it around.

My worry is that we can't plan for all this.  I can't really think of ways they won't work together (except minimizing and maximizing the same thing), so I'm not as worried about that, but naming them is going to be hard, and we need to get reuse as the default soon, so that we can switch to full hash and fix the current somewhat thrashy CI situation.  E4S, it turns out, has around 30 builds that are in conflict with some other build -- in the sense that they share a DAG hash but not a full hash.  We need that deployment model to start matching up -- we can't ignore build deps in our installs anymore, but I don't want to inflict that on users without default `--reuse`.

I would argue that `--reuse` and `--fresh` are special enough that they could have special options.  Can we move the "how many kinds of optimization are there" discussion to #28470 and work on getting this change in so that we can make `--reuse` default sooner?

I think if we do that, the only decision left here is "what do we call `--fresh`.  I have some opinions on that:

* I don't think `--deterministic` is good because people will use it on `develop` and they'll say "it's not deterministic".
* I don't think `--reproducible` is good because we still don't do fully reproducible builds in Spack -- the environment can matter.
* I think `--new` is about as good as `--fresh`, but @becker33 didn't like `--new` for the reasons mentioned above.
* I have really started liking `--fresh` *because* we can define it.  I don't think Spack's default install mode right now is much like anything else.

One of the reasons I like `--fresh` is because it maps to a lot of use cases:
* In CI, it means we're testing the latest configuration.  
* In an env or in regular Spack, it means we're upgrading -- but not just versions.
* I think `--latest` could be good but it sort of implies just versions. 
* I think it works well in conversation, especially with a link to a precise definition in the docs.

Do you have other suggestions?

**Comment by adamjstewart** [comment]
Yep, I definitely agree we should make `--reuse` the default soon, just trying to help with the naming discussion. Like you said, we don't want to have to support multiple naming schemes at the same time, so it's better to get this right the first time.

> I can't really think of ways they won't work together (except minimizing and maximizing the same thing), so I'm not as worried about that

What about "reuse all" vs "reuse deps" vs "reuse none"? That set alone makes me wonder whether boolean key-value pairs make sense. We could have separate `reuse: true/false, deps_only: true/false` settings, but that wouldn't work if you want to reuse deps but maximize root spec versions. I think we're going to need a more complicate naming scheme than the one you're proposing here.

Another alternative to `--fresh`: `--no-reuse`. It isn't clear that `--fresh` and `--reuse` are opposites and mutually exclusive.

**Comment by scottwittenburg** [comment]
> It isn't clear that `--fresh` and `--reuse` are opposites and mutually exclusive.

I agree it's not intuitive that these are mutually exclusive, and that's the main thing I don't like about `--fresh`.  I think I prefer `--no-reuse`, but based on the description it seems it might not be accurate.  The current behavior is basically `--fresh` and we already get *some* reuse from that, so `--no-reuse` is slightly misleading.  Is that the main argument against it?

**Comment by nilsvu** [comment]
For what it's worth, I like `--fresh`. It captures the user's intention pretty well (plus the more thought-out reasons given by @tgamblin). Whichever name you end up choosing, I would like to suggest adding a `-U` shorthand for the option, since it often _has_ upgrade semantics and users are familiar with this option from package managers like `pip`.

Also, @adamjstewart I love your suggestion for a `pure chaos` config option. It would be super useful for CI testing.

**Comment by alalazo** [comment]
I concur that `--update`, `--new`, `--deterministic` and `--reproducible` are all imprecise for the reasons explained in https://github.com/spack/spack/pull/28468#issuecomment-1015060972. The `--no-reuse` option is not bad, but it is also imprecise since we could reuse on a hash match. Would that work if we turned it into the slightly longer `--disregard-reuse`?

**Comment by alalazo** [comment]
All in all, I like both `--no-reuse` (or slight variation on the same theme) or `--fresh`.

**Comment by nilsvu** [comment]
Another suggestion: go with @adamjstewart's suggestion above right away(https://github.com/spack/spack/pull/28468#issuecomment-1015035473) and make the CLI option something like `--reuse=none` or `--reuse=all`.

**Comment by ashermancinelli** [comment]
+1 for `--no-reuse` but a poll in the Spack slack channel seems in order. The full `--concretization-strategy=foo` is ultimately what I want, and would be extremely helpful for CI as has already been mentioned.

**Comment by aweits** [comment]
`--deprioritize-reuse` is a thought that just occurred to me...

**Comment by tgamblin** [comment]
I kicked off the poll with `--no-reuse` and `--fresh`, as we got a few votes for each of these here and on the call this morning (and I'm trying to keep it simple).  Vote in Slack!

**Comment by tgamblin** [comment]
@adamjstewart::

> What about "reuse all" vs "reuse deps" vs "reuse none"?

I think we can handle cases like this with, e.g.:

Now:
```yaml
concretizer:
    reuse: true
```

Later:
```yaml
concretizer:
    reuse:
        from: [env, global, caches]
        only: [deps]
```

`reuse: true` would still imply a default configuration.  We did this with `view: true` vs. `view:` with a full projection specification in environments already -- I think it's relatively easy to do this for concretizer prefs, too.

I know I said it was hard to transition from a string to a dict above, but it's fairly easy to do when you're *expanding* on the `true` case.  `config:concretizer` is harder to transition this way b/c we want to deprecate `original` and eventually move it out of `config`.  Similar for `concretization: together` in the env -- the semantics don't telescope as naturally.  I think it can work at least for `reuse`.

**Comment by tgamblin** [comment]
@adamjstewart: ok based on the Slack poll, I think `--fresh` won.

I think the last thing that we might need here is docs.  Anything else missing in your view?

Screen Shot 2022-01-20 at 2 08 15 PM

**Comment by haampie** [comment]
Wow, did not expect people would go for --fresh.

**Comment by tgamblin** [comment]
I added a `-U` short option for `--fresh` per @nilsleiffischer's recommendation.  I'll update the docs tomorrow...

File: lib/spack/spack/environment/environment.py [hunk 1]
```diff
@@ -1160,7 +1160,7 @@ def _concretize_together(self, tests=False, reuse=False):
             self._add_concrete_spec(abstract, concrete)
         return concretized_specs
 
-    def _concretize_separately(self, tests=False, reuse=False):
+    def _concretize_separately(self, tests=False, reuse=None):
```
Comment (becker33) [inline lib/spack/spack/environment/environment.py]: It looks like this argument isn't used

File: lib/spack/spack/schema/concretizer.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+.. literalinclude:: _spack_root/lib/spack/spack/schema/concretizer.py
+   :lines: 13-
+"""
+
+properties = {
+    'concretizer': {
+        'type': 'object',
+        'additionalProperties': False,
+        'properties': {
+            'reuse': {'type': 'boolean'},
```
Comment (becker33) [inline lib/spack/spack/schema/concretizer.py:17]: Shouldn't the `original|clingo` choice move here as well? And maybe `together|separately` too

File: lib/spack/spack/solver/asp.py [hunk 3]
```diff
@@ -2026,33 +2032,74 @@ def _develop_specs_from_env(spec, env):
...
-        driver.out = sys.stdout
-
-    # Check upfront that the variants are admissible
-    for root in specs:
-        for s in root.traverse():
-            if s.virtual:
-                continue
-            spack.spec.Spec.ensure_valid_variants(s)
+    def __init__(self):
+        self.set_default_coanfiguration()
```
Comment (becker33) [inline lib/spack/spack/solver/asp.py]:
```suggestion
        self.set_default_configuration()
```

File: lib/spack/spack/solver/asp.py [hunk 4]
```diff
@@ -2026,33 +2032,74 @@ def _develop_specs_from_env(spec, env):
...
+            for s in root.traverse():
+                if s.virtual:
+                    continue
+                spack.spec.Spec.ensure_valid_variants(s)
 
-    setup = SpackSolverSetup()
-    return driver.solve(
-        setup, specs, dump, models, timers, stats, tests, reuse
-    )
+        if self.reuse is None:
```
Comment (becker33) [inline lib/spack/spack/solver/asp.py]: Why preserve a reuse option in the code in addition to the config? It doesn't look like the reuse options on any of these methods are actually hooked up to any code entrypoints, and I don't see them being used in testing either.

File: lib/spack/spack/solver/asp.py [hunk 5]
```diff
@@ -644,7 +644,7 @@ def stringify(x):
 class SpackSolverSetup(object):
     """Class to set up and run a Spack concretization solve."""
 
-    def __init__(self):
+    def __init__(self, reuse=False, tests=False):
```
Comment (becker33) [inline lib/spack/spack/solver/asp.py:665]: This is the only `reuse` method argument that seems to be used for anything

File: lib/spack/spack/solver/asp.py [hunk 6]
```diff
@@ -2026,33 +2032,74 @@ def _develop_specs_from_env(spec, env):
...
+        # These properties are settable via spack configuration. `None`
+        # means go with the configuration setting; user can override.
+        self.reuse = None
+
+        # these are concretizer settings
+        self.dump = ()
+        self.models = 0
+        self.timers = False
+        self.stats = False
+        self.tests = False
```
Comment (becker33) [inline lib/spack/spack/solver/asp.py]: Any reason these have to be set separately instead of either as constructor args or as solve args?

File: lib/spack/spack/cmd/common/arguments.py [hunk 7]
```diff
@@ -322,11 +322,68 @@ def add_cdash_args(subparser, add_help):
...
+
+        super(ConfigSetAction, self).__init__(
+            option_strings=option_strings,
+            dest=dest,
+            nargs=0,
+            const=const,
+            default=default,
+            required=required,
+            help=help
+        )
```
Comment (becker33) [inline lib/spack/spack/cmd/common/arguments.py:355]:
```suggestion
    def __init__(self,
                 option_strings,
                 config_path,
                 const,
                 default=None,
                 required=False,
                 help=None,
                 metavar=None):
        # save the config option we're supposed to set
        self.config_path = config_path

        super(ConfigSetAction, self).__init__(
            option_strings=option_strings,
            dest='fake',
            nargs=0,
            const=const,
            default=default,
            required=required,
            help=help
        )
```

I think this might be more legible to make clear we aren't using `dest`.

**Review by becker33 — CHANGES_REQUESTED** [review]
Two comments on the new Solver interface
1.  I think the attributes would be more easily handled as args, either to the constructor or the solve method
2. I think we should fully rely on the config system for the reuse options, instead of having both. We aren't using the method interface for it, and I think leaving that around will confuse maintainers.

File: lib/spack/spack/schema/concretizer.py (hunk 2 above)
Comment (tgamblin) [reply lib/spack/spack/schema/concretizer.py:17]: I think the original|clingo choice should just go away when we're done with it, so I think this goes in a follow-on PR.  

I agree about together/separately, but I think we should move that when we add "as together as possible" and not do it here.

File: lib/spack/spack/solver/asp.py (hunk 6 above)
Comment (tgamblin) [reply lib/spack/spack/solver/asp.py]: basically I think there are going to be too many of these, and doing this makes it less unwieldy to set most of them from config and allow the user to override defaults when they want to.  What do you think?

File: lib/spack/spack/cmd/common/arguments.py (hunk 7 above)
Comment (tgamblin) [reply lib/spack/spack/cmd/common/arguments.py:355]: I disagree w/this b/c now we have a variable called `fake` that nobody can query, because what's in it depends on how many of these config args you add.  The last one will win.  Having the variable saved into a unique location -- so that you can query later *whether* it was set from the command line -- is something users of this class may want to do.

Comment (tgamblin) [reply lib/spack/spack/cmd/common/arguments.py:355]: It's normal for argparse arguments to be available on the namespace that argparse produces - I don't see why this should be different (or handicapped).

File: lib/spack/spack/solver/asp.py (hunk 4 above)
Comment (tgamblin) [reply lib/spack/spack/solver/asp.py]: This is a config property that users can set (which incidentally is how clingo manages this internally, b/c it has tons of options).  Generally when you have a *lot* of options, it's better to let the user modify a template (the defaults) than duplicate everything in every method signature.  I'm trying to avoid having these repeated everywhere in the code -- to reduce the places we have to modify to add new options.

File: lib/spack/spack/environment/environment.py (hunk 1 above)
Comment (tgamblin) [reply lib/spack/spack/environment/environment.py]: It's used here: https://github.com/spack/spack/pull/28468/files#diff-9b1efe5b325365c46244290e635d28b948c1f0b40a54432468bc23ce422aece8L1188

**Comment by tgamblin** [comment]
@becker33 @adamjstewart I've added docs for this and commented on @becker33's feedback.  Can you take another look?

File: lib/spack/spack/solver/asp.py (hunk 6 above)
Comment (becker33) [reply lib/spack/spack/solver/asp.py]: I think they should still be available to the constructer as kwargs. Something like

```
def __init__(self, **kwargs):
    configuration = default_configuration.copy()
    configuration.update(kwargs)
    self.set_configuration(configuration)
```

I think that makes for an interface that will make more sense to folks using this object in the future.

File: lib/spack/spack/solver/asp.py (hunk 4 above)
Comment (becker33) [reply lib/spack/spack/solver/asp.py]: I think that the default should be pulled from config at `Solver` instantiation time, rather than in `Solver.solve`. If we use my `configuration_defaults` code from above, that dict could include `reuse: spack.config.get('concretizer:reuse', False')`

**Review by becker33 — CHANGES_REQUESTED** [review]
I think these are my only two outstanding requests

File: lib/spack/spack/schema/concretizer.py (hunk 2 above)
Comment (becker33) [reply lib/spack/spack/schema/concretizer.py:17]: I'm okay with leaving that for the next PR.

File: lib/spack/spack/solver/asp.py [hunk 8]
```diff
@@ -2016,33 +2022,74 @@ def _develop_specs_from_env(spec, env):
...
-#
-def solve(specs, dump=(), models=0, timers=False, stats=False, tests=False,
-          reuse=False):
-    """Solve for a stable model of specs.
-
-    Arguments:
-        specs (list): list of Specs to solve.
-        dump (tuple): what to dump
-        models (int): number of models to search (default: 0)
+class Solver(object):
```
Comment (becker33) [inline lib/spack/spack/solver/asp.py:2047]: Thoughts on this class after our webex meeting:

The `Solver` class constructor should contain any calls to `spack.config` to get concretizer defaults -- any changes to the config should not propagate to an existing object. Client code (`Spec.concretize`, `Environment.concretize`) should instantiate a new `Solver` object if it needs a new one, or it can explicitly set attributes to control the behavior of an existing `Solver` object.

The `Solver` object should also come with a driver. For now there's only one driver, but we should instantiate a `PyclingoDriver` at `Solver` instantiation time.

The `out` argument for `PyclingoDriver` should move to the `solve` method. All output-related arguments (`show`, `models`, `timers`, `stats`) should be arguments to the solve method, because they're relevant to a specific solve, not to the solver object.

The `SolverSetup` object has to be specific to a given solve, because it maintains state. So `Solver.solve` instantiates a `SolverSetup` and calls `driver.solve` with that setup object.

I also think the driver should become a required constructor parameter for the `SolverSetup` object -- we're never going to use the same SolverSetup object with different drivers.

Does that all make sense and allow for the simplification of the interface that you were looking for with the changes here? Trying to go with a rules-based description rather than code suggestions so the logic behind it is clearer, and whatever we agree on we can put into the docstrings.

**Review by becker33 — CHANGES_REQUESTED** [review]

File: lib/spack/spack/solver/asp.py (hunk 8 above)
Comment (tgamblin) [reply lib/spack/spack/solver/asp.py:2047]: I did all of these *except* for this one:

> I also think the driver should become a required constructor parameter for the `SolverSetup` object -- we're never going to use the same `SolverSetup` object with different drivers.

Mainly because the `SolverSetup` can be used *without* a driver -- as it is in `spack diff` to create facts for diffing.  I don't think we should construct a driver there and couple the two classes that tightly, when all we want is to get facts for a set of specs.

File: lib/spack/spack/solver/asp.py [hunk 9]
```diff
@@ -527,10 +525,22 @@ def fact(self, head, assumption=False):
...
-            timers=False, stats=False, tests=False, reuse=False,
-    ):
+    def solve(self, setup, specs, nmodels=0, timers=False, stats=False, out=None):
+        """Set up the input and solve for dependencies of ``specs``.
+
+        Arguments:
+          setup (SpackSolverSetup): An object to set up the ASP problem.
+          specs (list): List of ``Spec`` objects to solve for.
+          nmodels (list): Number of models to consider (default 0 for unlimited).
+          timers (bool):  Print out coarse fimers for different solve phases.
```
Comment (becker33) [inline lib/spack/spack/solver/asp.py]:
```suggestion
          timers (bool):  Print out coarse timers for different solve phases.
```

File: lib/spack/spack/solver/asp.py [hunk 10]
```diff
@@ -2016,33 +2032,60 @@ def _develop_specs_from_env(spec, env):
...
+        """
+        Arguments:
+          specs (list): List of ``Spec`` objects to solve for.
+          show (tuple): What information to print to the console while running.
+            Options are:
+            * asp: asp program text
+            * opt: optimization criteria for best model
+            * output: raw clingo output
+            * solutions: models found by asp program
+            * all: all of the above
```
Comment (becker33) [inline lib/spack/spack/solver/asp.py]: Is `all` still an option at this point? I think we parse `all` in the command and it's the lists of the others at this poinit

File: lib/spack/spack/solver/asp.py [hunk 11]
```diff
@@ -2016,33 +2032,60 @@ def _develop_specs_from_env(spec, env):
...
+        """
+        # Check upfront that the variants are admissible
+        for root in specs:
+            for s in root.traverse():
+                if s.virtual:
+                    continue
+                spack.spec.Spec.ensure_valid_variants(s)
+
+        setup = SpackSolverSetup(reuse=self.reuse, tests=tests)
+        return self.driver.solve(
```
Comment (becker33) [inline lib/spack/spack/solver/asp.py:2096]: The driver might need access to the `show` argument later to fix a completely separate bug where `solve --show asp` fails when the solve fails, even though we didn't ask for the solutions.

File: lib/spack/spack/test/concretize.py [hunk 12]
```diff
@@ -1354,7 +1354,8 @@ def test_concrete_specs_are_not_modified_on_reuse(
         # when reused specs are added to the mix. This prevents things
         # like additional constraints being added to concrete specs in
         # the answer set produced by clingo.
-        s = spack.spec.Spec(spec_str).concretized(reuse=True)
+        with spack.config.override("concretizer:reuse", True):
```
Comment (becker33) [inline lib/spack/spack/test/concretize.py:1357]: This test now needs the `config` fixture

**Review by becker33 — CHANGES_REQUESTED** [review]

File: lib/spack/spack/solver/asp.py (hunk 10 above)
Comment (tgamblin) [reply lib/spack/spack/solver/asp.py]: actually none of these are used by `asp.py`. I'll consolidate it to just output.

File: lib/spack/spack/test/concretize.py (hunk 12 above)
Comment (tgamblin) [reply lib/spack/spack/test/concretize.py:1357]: yep thanks

File: lib/spack/spack/cmd/solve.py [hunk 13]
```diff
@@ -102,21 +104,29 @@ def solve(parser, args):
...
+    # Note: reuse and other concretizer prefs are passed as configuration
+    solver = asp.Solver()
+    output = sys.stdout if "asp" in show else None
+    result = solver.solve(
+        specs,
+        out=output,
+        models=models,
+        timers=args.timers,
+        stats=args.stats,
+        setup_only=not any(s in show for s in ("output", "solutions"))
```
Comment (becker33) [inline lib/spack/spack/cmd/solve.py]: you need to run the solve to show `opt` too

```suggestion
        setup_only=(set(show) != {'asp'})
```

File: lib/spack/spack/cmd/solve.py [hunk 14]
```diff
@@ -102,21 +104,29 @@ def solve(parser, args):
...
+    if 'solutions' not in show:
         return
 
     # die if no solution was found
     result.raise_if_unsat()
 
-    # dump the solutions as concretized specs
-    if 'solutions' in dump:
+    # show the solutions as concretized specs
+    if 'solutions' in show:
```
Comment (becker33) [inline lib/spack/spack/cmd/solve.py:126]: `opt` won't work without solutions still, but that can be a separate PR to fix it.

**Review by becker33 — CHANGES_REQUESTED** [review]

**Review by becker33 — DISMISSED** [review]

**Review by becker33 — DISMISSED** [review]

**Review by becker33 — APPROVED** [review]

File: lib/spack/spack/solver/asp.py [hunk 15]
```diff
@@ -545,9 +563,13 @@ def solve(
...
         with self.control.backend() as backend:
             self.backend = backend
-            solver_setup.setup(self, specs, tests=tests, reuse=reuse)
+            setup.setup(self, specs)
         timer.phase("setup")
 
+        # If we're only doing setup, just return an empty solve result
+        if setup_only:
+            return Result(specs)
+
```
Comment (alalazo) [inline lib/spack/spack/solver/asp.py:572]: I think this is wrongly positioned, we should exit _after_ we generated the error facts. I'll submit a bug fix asap.
